python/GafferUI/PlugPopup.py

```diff
@@ -95,9 +85,113 @@ def __keyPress( self, widget, event ) :
 		if event.key == "Return" :
 			self.close()
 
+class PlugPopup( _PopupWindow ) :
+
+	def __init__( self, plugs, title = None, warning = None, **kw ) :
+
+		assert( len( plugs ) )
+		script = sole( p.ancestor( Gaffer.ScriptNode ) for p in plugs )
+		assert( script is not None )
+
+		column = GafferUI.ListContainer( spacing = 4 )
+		_PopupWindow.__init__( self, "", child = column, **kw )
+
+		with column :
+
+			# Title
+
+			if title is None :
+
+				# Make default title
+
+				if len( plugs ) == 1 :
+					title = plugs[0].relativeName( script )
+				else :
+					nodes = { plug.node() for plug in plugs }
+					title = "{} plugs{}".format(
+						len( plugs ),
+						" on {} nodes".format( len( nodes ) ) if len( nodes ) > 1 else ""
+					)
+
+					commonNode = plugs[0].node()
+					for plug in plugs :
+						if not commonNode.isAncestorOf( plug ) :
+							commonNode = commonNode.commonAncestor( plug.node() )
+
+					if commonNode != script :
+						title = "{} ({})".format(
+							commonNode.relativeName( script ),
+							title
+						)
+
+			if title :
+
+				titleWidget = GafferUI.Label( "<h4>{}</h4>".format( title ) )
+				titleWidget.setToolTip(
+					"\n".join( "- " + p.relativeName( script ) for p in plugs )
+				)
+
+			# Widget row
+
+			with GafferUI.ListContainer( spacing = 4, orientation = GafferUI.ListContainer.Orientation.Horizontal ) :
+
+				# Warning
+
+				if warning :
+					warningBadge = GafferUI.Image( "warningSmall.png" )
+					warningBadge.setToolTip( warning )
+
+				# PlugValueWidget, or label explaining why we can't show one
+
+				if sole( p.__class__ for p in plugs ) :
+					self.__plugValueWidget = GafferUI.PlugValueWidget.create( plugs )
+				else :
+					self.__plugValueWidget = None
+					GafferUI.Label( "Unable to edit plugs with mixed types" )
+
+		self.visibilityChangedSignal().connect( Gaffer.WeakMethod( self.__visibilityChanged ), scoped = False )
+
+	def popup( self, center = None ) :
+
+		_PopupWindow.popup( self, center )
+
+		# Attempt to focus the first text widget. This is done after making
+		# the window visible, as we check child widget visibility to avoid
+		# attempting to focus hidden widgets.
+
+		textWidget = self.__textWidget( self.__plugValueWidget )
+		if textWidget is not None :
+			if isinstance( textWidget, GafferUI.TextWidget ) :
+				textWidget.grabFocus()
+				textWidget.setSelection( 0, len( textWidget.getText() ) )
+			else :
+				textWidget.setFocussed( True )
+			textWidget._qtWidget().activateWindow()
+
+	def plugValueWidget( self ) :
+
+		return self.__plugValueWidget
+
+	def __visibilityChanged( self, unused ) :
+
+		if self.visible() :
+			return
+
+		# The modal popup will hide the window without de-focusing the
+		# current widget. Ultimately this results in un-committed edits
+		# from being lost. Force the active widget to resign focus.
```

Comment:
The word "from" doesn't fit here.  But perhaps more to the point, I don't really understand this at all.

Trying to work out a guess as to what's happening:  we think that the only way we can lose visibility is if a modal popup happens?  We don't really interact at all with this hypothetical modal popup in this code, but it's the reason why we need to catch a change in visibility?

The focusWidget we get in the code below is a child widget of ours ( ie. the actual text field ).  At the moment when this runs, it's still the focusWidget, even though I assume it's about to be unfocussed ( it's a little weird that `self.visible()` is already updated, but focusWidget isn't updated yet )?

We then manually tell this widget to defocus - I guess it's effectively a Qt bug that hiding something doesn't count as defocusing it?

Reply:
I have to confess that I don't know the details here either - this was stuff that Tom did. Of the two things I was refactoring into one, only the SceneViewInspector had this workaround, while the Spreadsheet seems to have survived without it. In the absence of understanding and an original repro, I thought it safer to keep the workaround. I'll update the comment to reflect my limited understanding a bit better...

Reply:
Attempted clarification in d8c560b.